src/image/read/specific_channels.rs

```diff
@@ -279,30 +280,85 @@ pub struct OptionalSampleReader<DefaultSample> {
 impl<Sample: FromNativeSample> SampleReader<Sample> {
     fn read_own_samples<'s, FullPixel>(
         &self, bytes: &'s[u8], pixels: &mut [FullPixel],
-        get_pixel: impl Fn(&mut FullPixel) -> &mut Sample
+        get_sample: impl Fn(&mut FullPixel) -> &mut Sample
     ){
         let start_index = pixels.len() * self.channel_byte_offset;
         let byte_count = pixels.len() * self.channel.sample_type.bytes_per_sample();
-        let mut own_bytes_reader = &bytes[start_index .. start_index + byte_count]; // TODO check block size somewhere
-
-        let error_msg = "error when reading from in-memory slice";
+        let mut own_bytes_reader = &mut &bytes[start_index .. start_index + byte_count]; // TODO check block size somewhere
+        let output = pixels.iter_mut().map(|pixel| get_sample(pixel));
 
         // match outside the loop to avoid matching on every single sample
         match self.channel.sample_type {
-            SampleType::F16 => for pixel in pixels.iter_mut() {
-                *get_pixel(pixel) = Sample::from_f16(f16::read(&mut own_bytes_reader).expect(error_msg));
-            },
-
-            SampleType::F32 => for pixel in pixels.iter_mut() {
-                *get_pixel(pixel) = Sample::from_f32(f32::read(&mut own_bytes_reader).expect(error_msg));
-            },
-
-            SampleType::U32 => for pixel in pixels.iter_mut() {
-                *get_pixel(pixel) = Sample::from_u32(u32::read(&mut own_bytes_reader).expect(error_msg));
-            },
+            SampleType::F16 => read_and_convert_samples_batched(
+                &mut own_bytes_reader, output,
+                Sample::from_f16s
+            ),
+
+            SampleType::F32 => read_and_convert_samples_batched(
+                &mut own_bytes_reader, output,
+                Sample::from_f32s
+            ),
+
+            SampleType::U32 => read_and_convert_samples_batched(
+                &mut own_bytes_reader, output,
+                Sample::from_u32s
+            ),
         }
 
         debug_assert!(own_bytes_reader.is_empty(), "bytes left after reading all samples");
+
+
+        /// performs something similar to
+        /// `for sample in out_samples { *sample = Sample::convert_from(f16/f32/u32::read_from_bytes(bytes)); }`
+        fn read_and_convert_samples_batched<'t, From, To>(
+            mut bytes: impl Read,
+            mut out_samples: impl ExactSizeIterator<Item=&'t mut To>,
+            convert_slice: impl Fn(&[From], &mut [To])
+        ) where From: Data + Default + Copy, To: 't + Default + Copy
+        {
+            // using a batch size of 4
+            // because that's what `half` has vectorization for,
+            // and we want the compiler to
+            // optimize away all the logic in
+            // `HalfFloatSliceExt::convert_from_f32_slice`
```

Comment:
`half` supports converting 8 values at once in the latest git: https://github.com/starkat99/half-rs/issues/66 (not released to crates.io yet)

AVX-512 brings conversions of 16 values at once, which `half` also intends to support: https://github.com/starkat99/half-rs/issues/62

Also, right now `half` performs feature detection for each slice of 4 values, but the latest git performs it once per slice: https://github.com/starkat99/half-rs/commit/bc38c5b237e822e187cc1ac710b1099cc2faafd3
So bigger slice sizes are beneficial because the check only needs to be done once per the whole big slice instead of once per chunk of 4 or 8.

This needs to be balanced against not using too much CPU cache. I suggest the length of 16 for a start, since that's the minimum length that can take advantage of an AVX-512 instruction.

Reply:
Can we somehow avoid having our own batching logic, and inside each batch still have the batching logic from `half`? I imagine it would be rather duplicate and lead to some performance problems

Reply:
That's actually a good thing! First we slice the input into batches of a size we choose that fits into the CPU cache, and `half` can then perform feature detection once per large slice (instead of once per small batch) and convert numbers in the most efficient way available in hardware, with a batch size we don't have to know or care about.

`chunks_exact_mut` that the `half` loop uses is cheap, and with a fixed-size slice (like we have) it will be cheaper still. Even the logic handling the remainder should be simply removed by the compiler, since the slice method in `half` is `#[inline]` and the batch size is constant.

Reply:
that's just what i hoped for, great :)